[JSC] Refactor Loop Unrolling to Use DFGCloneHelper

hyjorc1 opened the pull request:
#### 91b72777aecbc20988f1ecaab5bd38d501c9da71
<pre>
[JSC] Refactor Loop Unrolling to Use DFGCloneHelper
<a href="https://bugs.webkit.org/show_bug.cgi?id=290520">https://bugs.webkit.org/show_bug.cgi?id=290520</a>
<a href="https://rdar.apple.com/148001458">rdar://148001458</a>

Reviewed by Justin Michaud.

This patch:
1. Introduced DFGCloneHelper to encapsulate node cloning logic
   for the loop unrolling phase.
2. Moved isNodeCloneable, cloneNode, and cloneNodeImpl from DFGLoopUnrollingPhase.cpp
   into the new helper class.
3. Simplifies and modularizes the unrolling phase, improving maintainability
   and separation of concerns.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp: Added.
(JSC::DFG::LoopUnrollingCloneHelper::isNodeCloneable):
(JSC::DFG::LoopUnrollingCloneHelper::cloneNode):
(JSC::DFG::LoopUnrollingCloneHelper::cloneNodeImpl):
* Source/JavaScriptCore/dfg/DFGCloneHelper.h: Added.
(JSC::DFG::LoopUnrollingCloneHelper::LoopUnrollingCloneHelper):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::canCloneLoop):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::LoopUnrollingPhase::isNodeCloneable): Deleted.
(JSC::DFG::LoopUnrollingPhase::cloneNode): Deleted.
(JSC::DFG::LoopUnrollingPhase::cloneNodeImpl): Deleted.

Canonical link: <a href="https://commits.webkit.org/292839@main">https://commits.webkit.org/292839@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8e470a6ba150ffc8e06c0015638d3ba9301255a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16717 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6925 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102175 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47619 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17009 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25167 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74092 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31177 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100097 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87814 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54306 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12591 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5677 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46949 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89767 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82654 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5754 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104197 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95715 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24168 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17630 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83013 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24545 "Failed to checkout and rebase branch from PR 43099") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83928 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82414 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20772 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4649 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17673 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24133 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29284 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119342 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23956 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33522 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25529 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingCloneHelper.cpp:
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
Will change `2024` to `2025`.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingCloneHelper.h:
```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
Will change `2024` to `2025`.

justinmichaud reviewed (approved):
What do you think about naming it just DFGCloneHelper? I could imagine some other passes might want to clone nodes too

hyjorc1 committed:
[JSC] Refactor Loop Unrolling to Use DFGCloneHelper
https://bugs.webkit.org/show_bug.cgi?id=290520
rdar://148001458

Reviewed by Justin Michaud.

This patch:
1. Introduced DFGCloneHelper to encapsulate node cloning logic
   for the loop unrolling phase.
2. Moved isNodeCloneable, cloneNode, and cloneNodeImpl from DFGLoopUnrollingPhase.cpp
   into the new helper class.
3. Simplifies and modularizes the unrolling phase, improving maintainability
   and separation of concerns.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp: Added.
(JSC::DFG::LoopUnrollingCloneHelper::isNodeCloneable):
(JSC::DFG::LoopUnrollingCloneHelper::cloneNode):
(JSC::DFG::LoopUnrollingCloneHelper::cloneNodeImpl):
* Source/JavaScriptCore/dfg/DFGCloneHelper.h: Added.
(JSC::DFG::LoopUnrollingCloneHelper::LoopUnrollingCloneHelper):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::canCloneLoop):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::LoopUnrollingPhase::isNodeCloneable): Deleted.
(JSC::DFG::LoopUnrollingPhase::cloneNode): Deleted.
(JSC::DFG::LoopUnrollingPhase::cloneNodeImpl): Deleted.

Canonical link: https://commits.webkit.org/292839@main
